common/sampling.cpp

```diff
@@ -284,6 +287,9 @@ static llama_token_data_array llama_sampling_prepare_impl(
 
     // Get a pointer to the logits
     float * logits = llama_get_logits_ith(ctx_main, idx);
+    if (!logits) {
+        throw std::runtime_error("llama_get_logits_ith failed");
```

Comment:
>  This change potentially helps prevent the server from crashing.

This exception (like the others added) is never caught, so it will still make the server crash. With `bge-small` and trying to make a chat completion:

```
terminate called after throwing an instance of 'std::runtime_error'
  what():  llama_get_logits_ith failed
Aborted (core dumped)
```

Maybe an error code could be returned (using negative numbers, assuming no `llama_token` is negative...), so that the error can then be handled at each call site?

And if it's not handled, maybe make `llama_sampling_accept` crash when it gets a negative `llama_token`?

Or alternatively, slot processing in the server could be wrapped in a `try` block?

I'm not sure what's the cleanest solution here.

Reply:
Doesn't the server have a per request exception handler? Worst case is the client request crashes and 500's, no?

Reply:
I tested this before writing, and the server crashes with the error I quoted. On the client side, this is what happens:
```
curl: (18) transfer closed with outstanding read data remaining
```
> Doesn't the server have a per request exception handler?

I would have thought so too, but it might only handle the exceptions of the http library, I'm not sure.

https://github.com/ggerganov/llama.cpp/blob/201cc11afa0a1950e1f632390b2ac6c937a0d8f0/examples/server/server.cpp#L2988-L3001

Reply:
OK I've pushed some stuff I think should help resolve this.